crates/dbsp/src/operator/dynamic/balance/balancer.rs

```diff
@@ -0,0 +1,832 @@
+        let fixed_policy = if any_flushed && !all_flushed {
+            // println!("{} metadata: {:?}", Runtime::worker_index(), metadata);
+            // All workers must have the same policy.
+            assert!(metadata
+                .iter()
+                .all(|worker_metadata| worker_metadata.is_some()
+                    && worker_metadata.as_ref().unwrap().current_policy
+                        == metadata[0].as_ref().unwrap().current_policy));
+            Some(metadata[0].as_ref().unwrap().current_policy)
+        } else {
+            None
+        };
+
+        // println!("fixed_policy: {:?}, hints: {:?}", fixed_policy, hints);
+
+        if fixed_policy.is_some()
+            && hints.policy_hint.is_some()
+            && fixed_policy != hints.policy_hint
+        {
+            return Err(BalancerError::InvalidPolicyHint(
+                hints.policy_hint.unwrap(),
+                format!("the current policy {fixed_policy:?} can no longer be changed during the current transaction"),
+            ));
+        }
+
+        if hints.policy_hint.is_some() {
+            Ok(hints.policy_hint)
+        } else {
+            Ok(fixed_policy)
+        }
+    }
+
+    /// Compute combined key distribution from metadata collected from
+    /// RebalancingExchangeSender instances in all workers.
+    fn key_distribution(
+        metadata: &[Option<RebalancingExchangeSenderExchangeMetadata>],
+    ) -> KeyDistribution {
+        metadata.iter().fold(
+            KeyDistribution::new(Runtime::num_workers()),
+            |mut key_distribution, metadata| {
+                if let Some(metadata) = metadata {
+                    key_distribution.merge(&metadata.key_distribution);
+                };
+                key_distribution
+            },
+        )
+    }
+}
+
+#[derive(Default, Debug)]
+struct BalancerInner {
+    /// All integrals managed by the balancer:
+    /// origin node id -> (exchange sender node id, hints)
+    integrals: BTreeMap<NodeId, (NodeId, BalancerHints)>,
+
+    /// All joins managed by the balancer:
+    /// join node id -> (left input node id, right input node id, is left join)
+    joins: BTreeMap<NodeId, (NodeId, NodeId, bool)>,
+
+    /// Connected components of the join graph:
+    clusters: Vec<Cluster>,
+
+    /// Map from input stream to the cluster index it belongs to.
+    stream_to_cluster: BTreeMap<NodeId, usize>,
+
+    /// Circuit's metadata exchange used to collect metadata from RebalancingExchangeSender instances.
+    metadata_exchange: MetadataExchange,
+}
+
+impl BalancerInner {
+    fn new(metadata_exchange: &MetadataExchange) -> Self {
+        Self {
+            integrals: BTreeMap::new(),
+            joins: BTreeMap::new(),
+            clusters: Vec::new(),
+            stream_to_cluster: BTreeMap::new(),
+            metadata_exchange: metadata_exchange.clone(),
+        }
+    }
+
+    /// Compute an "optimal" policy for each stream in the cluster.
```

Comment:
Cluster is very overloaded, can this be called something else, e.g., balanced_region?